### PR TITLE
interfaces/wayland: on classic systems only consider the OS slot for auto-connect

### PR DESCRIPTION
**Pull request by AlanGriffiths**
interfaces/wayland: On classic systems only consider the OS slot for auto-connect.

Installing additional providers of the "wayland" interface should not inhibit auto-connection to the implicit slot.

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
Having multiple candidate providers causes auto-connect to fail with an INFO message hidden in a log file. E.g.

```plain
systemctl status -n 40 snapd.service | grep wayland
Sep 05 14:36:18 Octopull-desktop snapd[8881]: task.go:337: DEBUG: 2019-09-05T14:36:18+01:00 INFO cannot auto-connect plug mir-kiosk-apps:wayland, candidates found: egmde-confined-desktop:wayland, mir-kiosk:wayland, core:wayland
```

**Comment by anonymouse64** [comment]
I'm not sure that we should be fixing this for wayland specifically, I think that this is a larger problem in that if there are multiple candidate slots from different snaps, why should we auto-connect to the system one? I think the better thing is for the plug to be able to specify default-provider for the interface to use the system slot (much like how the content interface works), but I'm not sure if this has been considered for non-content interfaces before.

**Comment by AlanGriffiths** [comment]
> I'm not sure that we should be fixing this for wayland specifically, I think that this is a larger problem in that if there are multiple candidate slots from different snaps, why should we auto-connect to the system one? 

I'm not convinced that there is a general rule for all interfaces that are implicit on Classic, but I think this is right for the wayland interface. It is a reasonable expectation that a Wayland based snap installed on a Wayland based desktop works without additional configuration.

> I think the better thing is for the plug to be able to specify default-provider for the interface to use the system slot

How do you see that working on Core? On core there is no system slot. Autoconnecting to a snap provided slot (if there is exactly one) seems right.

File: interfaces/builtin/wayland.go [hunk 1]
```diff
@@ -157,9 +158,14 @@ func (iface *waylandInterface) UDevPermanentSlot(spec *udev.Specification, slot
 	return nil
 }
 
-func (iface *waylandInterface) AutoConnect(*snap.PlugInfo, *snap.SlotInfo) bool {
-	// allow what declarations allowed
-	return true
+func (iface *waylandInterface) AutoConnect(plug *snap.PlugInfo, slot *snap.SlotInfo) bool {
+  if release.OnClassic {
```
Comment (zyga) [inline interfaces/builtin/wayland.go]: You should probably `go fmt` this file.

File: interfaces/builtin/wayland.go [hunk 2]
```diff
@@ -157,9 +158,14 @@ func (iface *waylandInterface) UDevPermanentSlot(spec *udev.Specification, slot
...
+	  // allow connection to implicit slot
+    return slot.Snap.GetType() == snap.TypeOS
+  } else {
+    // allow what declarations allowed
+    return true
+  }
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: This is interesting because the base declaration is trying to handle this a bit:
```
  wayland:
    allow-installation:
      slot-snap-type:
        - app
        - core
    deny-connection:
      on-classic: false
    deny-auto-connection:
      on-classic: false
```

Which says to deny-auto-connection when not classic. Therefore, on classic distro, auto-connect. This is assuming that wayland on classic is coming from the host and not a snap. This probably should be adjusted to be:
```
  wayland:
    allow-installation:
      slot-snap-type:
        - app
        - core
    allow-connection:
      slot-snap-type:
      - core
    allow-auto-connection:
      slot-snap-type:
      - core
```
There are different ways of expressing this, but I think that this is the simplest. It also is a new variation of our documented patterns (see interfaces/policy/basedeclaration.go).

Comment (AlanGriffiths) [reply interfaces/builtin/wayland.go]: > ```
>   wayland:
>     allow-installation:
>       slot-snap-type:
>         - app
>         - core
>     deny-connection:
>       on-classic: false
>     deny-auto-connection:
>       on-classic: false
> ```
> 
> 
> Which says to deny-auto-connection when not classic. Therefore, on classic distro, auto-connect.

I find these incantations hard to follow, but I think you've read this wrong. (Either that, or the current behaviour is broken.)

At present, on both Core and Classic, if there's a single wayland slot then plugs auto-connect. That is, auto-connect is **not** denied on "not classic", it happens. (E.g. https://tutorials.ubuntu.com/tutorial/secure-ubuntu-kiosk#2)

> This probably should be adjusted to be:
> 
> ```
>   wayland:
>     allow-installation:
>       slot-snap-type:
>         - app
>         - core
>     allow-connection:
>       slot-snap-type:
>       - core
>     allow-auto-connection:
>       slot-snap-type:
>       - core
> ```
> 
> There are different ways of expressing this, but I think that this is the simplest. It also is a new variation of our documented patterns (see interfaces/policy/basedeclaration.go).

Maybe I'm misreading your proposal. Wouldn't this prevent auto-connection to mir-kiosk on Core?

I want a change so that if there's an implicit slot (as is the case on Classic) that this takes precedence.

File: interfaces/builtin/wayland.go (hunk 1 above)
Comment (AlanGriffiths) [reply interfaces/builtin/wayland.go]: Done

**Comment by AlanGriffiths** [comment]
I'm confused by this test failure:
```
FAIL: wayland_test.go:192: WaylandInterfaceSuite.TestAutoConnect

wayland_test.go:193:
    c.Assert(s.iface.AutoConnect(s.plugInfo, s.coreSlotInfo), Equals, true)
... obtained bool = false
... expected bool = true

OOPS: 1032 passed, 1 FAILED
--- FAIL: Test (0.73s)
```
Running locally, I see:
```
$ go test -check.f WaylandInterfaceSuite.TestAutoConnect
AppArmor status: apparmor is enabled and all features are available
OK: 0 passed
PASS
ok      github.com/snapcore/snapd/cmd/snapd     0.014s
```
How can I reproduce this failure?

File: interfaces/builtin/wayland.go (hunk 2 above)
Comment (jdstrand) [reply interfaces/builtin/wayland.go]: I misread the initial proposal. I see now in your response and in https://forum.snapcraft.io/t/the-wayland-interface-and-auto-connect/13063/1 what you are trying to address is when there are multiple providing slots, that the classic one be chosen rather than have snapd refuse to connect any. Sorry about that.

(My comments were only considering changing the behavior of a single providing slot)

Removing my comment since it is confusing.

**Comment by AlanGriffiths** [comment]
> I will say that for your specific use case, the request is reasonable to default to classic and it is easy to see why connecting an implicit classic slot would be preferable. However, I think it is also reasonable, generally speaking, for someone who installs a slot-providing snap to expect that it should be used instead of the classic implicit slot, so it is unclear to me what the default choice should be (ie, I still am leaning towards snapd's current implementation). Perhaps if there was a way to configure a preference of one over another, that could be useful.

I agree that in the general case there is no reasonable default that could be applied to all without further information from the interface and/or plug. It doesn't take much effort to suggest cases that benefit from a default suggested by the interface, the plug, or the publisher (and there's the slippery slope towards dynamic priority ordering).

For this specific interface, there is a reasonable default, and this provides it.

**Comment by pedronis** [comment]
@jdstrand I am a bit wary to do this change for one interface without considering what should be the policy for all the interface that allow either an implicit system slot or a slot from an app.

I have also a slightly preference if the change could be expressed in the base-declaration but maybe I'm confused about that.

**Comment by AlanGriffiths** [comment]
> @jdstrand I am a bit wary to do this change for one interface without considering what should be the policy for all the interface that allow either an implicit system slot or a slot from an app.

I don't believe there's any simple policy for all interfaces that allow allow either an implicit system slot or a slot from an app. It is only because of the way that that the wayland interface works that there is an obvious default.

> I have also a slightly preference if the change could be expressed in the base-declaration but maybe I'm confused about that.

I could find no way to express this in the base-declaration.

**Comment by pedronis** [comment]
Marking as blocked, this needs a step back for me @jdstrand and me to consider the overall picture of all interfaces in potentially similar situations. Without looking into details, I find it strange that we have exactly one interface where the change is desirable.

**Comment by AlanGriffiths** [comment]
> Without looking into details, I find it strange that we have exactly one interface where the change is desirable.

I'm not asserting that there is only one interface that ought to behave this way, only that this is one that does (and that the default behavior shouldn't change as there are others that should not).

With Wayland, auto-connecting to the system makes a lot of sense.  I don't see the harm in special casing this interface, as I don't think this is likely to be common.

The way the `wayland` interface works is also unusual as there's no real separation of slots: Whichever slot provider is connected, access is enabled to all providers. Selection of Wayland provider is mediated by the `WAYLAND_DISPLAY` environment variable (which is normal in the Wayland domain).

**Comment by pedronis** [comment]
closing until we have a more general discussion as mentioned in

https://github.com/snapcore/snapd/pull/8789#issuecomment-644591749

**Comment by AlanGriffiths** [comment]
> closing until we have a more general discussion as mentioned in
> 
> [#8789 (comment)](https://github.com/snapcore/snapd/pull/8789#issuecomment-644591749)

Is this need for a discussion being tracked as an issue somewhere? Where?

**Comment by pedronis** [comment]
> closing until we have a more general discussion as mentioned in
> #8789 (comment)
> Is this need for a discussion being tracked as an issue somewhere? Where?

I'll try to make sure we have a conversation around if not at the next midcycle

**Comment by wmww** [comment]
Any movement on this? Right now as far as I can tell Wayland snaps are basically broken unless you happen to know the specific command to fix them (for which there is no good path to discovery). Seems like we should make this work somehow.
